### PR TITLE
GitHub pull requests support

### PR DESCRIPTION
**Pull request by caktux**

### CODE REVIEW

File: master/buildbot/steps/vstudio.py [hunk 1]
```diff
@@ -415,7 +415,9 @@ def start(self):
         if self.platform is None:
             config.error('platform is mandatory. Please specify a string such as "Win32"')
 
-        command = ["%VCENV_BAT%",
```
Comment (tardyp) [inline master/buildbot/steps/vstudio.py]: looks like this does not belong to that PR

Comment (caktux) [inline master/buildbot/steps/vstudio.py]: Nope, but it fixes Windows builds

Comment (tardyp) [inline master/buildbot/steps/vstudio.py]: please put that in another PR, so that we can merge it faster. Looks like this one needs more tests, according to @sa2ajj

File: master/buildbot/status/web/hooks/github.py [hunk 2]
```diff
@@ -100,10 +101,10 @@ def handle_push(self, payload):
         user = None
         # user = payload['pusher']['name']
         repo = payload['repository']['name']
-        repo_url = payload['repository']['url']
+        repo_url = payload['repository']['clone_url']
         # NOTE: what would be a reasonable value for project?
         # project = request.args.get('project', [''])[0]
-        project = payload['repository']['full_name']
+        project = payload['repository']['name']
```
Comment (djmitche) [inline master/buildbot/status/web/hooks/github.py]: Do these represent changes in the Github push payload?  Or a behavior change?

Comment (caktux) [inline master/buildbot/status/web/hooks/github.py]: They were merely to match our codebase repositories and project names, which are using the `.git` urls instead of the web-facing urls (good practice I suppose) and the repo name instead of account+repo name of `full_name`. Codebases simply weren't matching because of this. This should probably be configurable instead.

Comment (djmitche) [inline master/buildbot/status/web/hooks/github.py]: It seems like a distinct change from the rest of the PR, and one that could cause compatibility problems.

Comment (caktux) [inline master/buildbot/status/web/hooks/github.py]: The current values are also quite arbitrary, if your codebases aren't using the web-facing urls instead of the actual git urls, and if project names aren't using the account/repo format...

**Comment by yarikoptic** [comment]
Somewhat useless 1c:  THANK YOU @caktux for the PR and @buildbot/developers for polishing it up!

**Comment by sa2ajj** [comment]
I can see a lot of my commits, probably this branch needs to be rebased

File: master/buildbot/status/web/hooks/github.py [hunk 3]
```diff
@@ -168,6 +184,73 @@ def _process_change(self, payload, user, repo, repo_url, project):
...
+
+        # refname = payload['pull_request']['head']['ref']
+        branch = 'refs/pull/%s/head' % number
+        category = 'pull-request'
+
+        if payload['pull_request']['mergeable'] == False:
+            log.msg("Pull request `%s' not mergeable, ignoring" % branch)
+            return changes
+
+        r = requests.get(payload['pull_request']['commits_url'] + "?per_page=100")
```
Comment (tardyp) [inline master/buildbot/status/web/hooks/github.py]: per_page=100 looks like a code smell.

Proper code needs to loop on all pages

Comment (tardyp) [inline master/buildbot/status/web/hooks/github.py]: Also needs to use txrequests. This might mean change the whole hook framework to be asynchronous

**Comment by tardyp** [comment]
I dont see any mocking of requests as well, so I doubt the tests are properly done.
Did you add test for pull-request messages?

**Comment by sa2ajj** [comment]
@caktux I understand that it might be a bit too much to ask :)  However, do you think you could split the changes into to PRs: one for hook and one for status?  These must be independent changes.

**Comment by sa2ajj** [comment]
And, on related note, if you feel you need help, please do not hesitate to request one.

**Comment by sa2ajj** [comment]
@caktux any updates? :)  This feature would be really great!

**Comment by caktux** [comment]
Well this works as is, the context in status is required for this to work properly with GitHub, so I'm not splitting the changes. I think it was @tardyp who said he'd work on using txrequests, then the 100 commits limit would need to be fixed, but tbh I don't have anything else to contribute to this PR

File: master/buildbot/status/web/hooks/github.py [hunk 4]
```diff
@@ -168,6 +184,73 @@ def _process_change(self, payload, user, repo, repo_url, project):
...
+            files = []
+            commit_url = ''
+            if 'url' in commit:
+                commit_url = commit['url']
+                r = requests.get(commit_url)
+                commit_files = json.loads(r.text)
+
+                if commit_files and 'files' in commit_files:
+                    for f in commit_files['files']:
+                        if f['status'] == 'added':
```
Comment (sa2ajj) [inline master/buildbot/status/web/hooks/github.py]: `if f['status'] in ('added', 'modified', 'removed'):` ?

File: master/buildbot/status/web/hooks/github.py [hunk 5]
```diff
@@ -168,6 +184,73 @@ def _process_change(self, payload, user, repo, repo_url, project):
 
         return changes
 
+    def _process_pull_request(self, payload, user, repo, repo_url, project, codebase=None):
```
Comment (sa2ajj) [inline master/buildbot/status/web/hooks/github.py]: if `user` is not used, why keep it as a parameter at all?

Comment (sa2ajj) [inline master/buildbot/status/web/hooks/github.py]: Since it's a private method, why `codebase` parameter is needed at all?  (It's available as `self._codebase`)

File: master/buildbot/status/web/hooks/github.py [hunk 6]
```diff
@@ -168,6 +184,73 @@ def _process_change(self, payload, user, repo, repo_url, project):
...
+        number = payload['number']
+
+        # We only care about opened/reopened/synchronize pull requests
+        if payload['action'] not in ['opened', 'reopened', 'synchronize']:
+            log.msg("Ignoring `%s': Not an opened pull request" % number)
+            return changes
+
+        # refname = payload['pull_request']['head']['ref']
+        branch = 'refs/pull/%s/head' % number
+        category = 'pull-request'
```
Comment (sa2ajj) [inline master/buildbot/status/web/hooks/github.py]: `cateogory` is not changed after this line, why introduce a new variable?

**Comment by tardyp** [comment]
I wont be able to work on it right away, and certainly not for eight

**Comment by sa2ajj** [comment]
@caktux what will be your plan in light of @tardyp's comment?

**Comment by caktux** [comment]
I really don't have time to work on this either for at least a few weeks, I have to let you guys take it from here.

**Comment by sa2ajj** [comment]
@caktux thanks for the clarification, let's see how we can take it from here...

**Comment by sa2ajj** [comment]
This PR is probably superceded by GH-1632

**Comment by sa2ajj** [comment]
Closing in favour of GH-1632
